Add InfluxDB v2 to v3 migration automation

trevorbonas opened the pull request:
*Issue #, if available:*

N/A.

*Description of changes:*

Adds the InfluxDB v2 to v3 migration script. The script is available standalone or included in an AMI, which can be built with Packer and deployed to an EC2 instance with Terraform.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

trevorbonas committed:
Add initial version of v2 to v3 migration

trevorbonas committed:
Add build directory and *.egg-info files to .gitignore

trevorbonas committed:
Solve parallel ingestion

trevorbonas committed:
Use Packer with Terraform for EC2 deployment

trevorbonas committed:
Make Python requirement more specific in README.md

trevorbonas committed:
Add initial version of integration tests

trevorbonas committed:
Use brand new line protocol test data instead of sample data

trevorbonas committed:
Reformat test_influxdb_v2_to_v3_migration.py

trevorbonas committed:
Update pydoc comments

trevorbonas committed:
Update README.md with testing instructions

trevorbonas committed:
Add instructions to install InfluxDB v2 CLI and daemon

trevorbonas committed:
Update Python file top-level comments

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/utils.py:
```diff
@@ -0,0 +1,48 @@
+import tarfile
+
+
+def get_secret(secret_name: str, region_name: str | None = None) -> dict[str, str]:
```
can we also have region configurable at the top-level script (and pass it into get_secret)? I didnt have the AWS_REGION env var set and it was looking in the wrong place. Or we should mention that AWS_REGION should be set in the README somewhere

trevorbonas replied:
I have added a `--region` argument to both scripts.

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
````diff
@@ -0,0 +1,157 @@
+       '{"INFLUXDB_V2_TOKEN": "replace me", "INFLUXDB_V3_TOKEN": "replace me"}'
+   ```
+6. [Download and Install the InfluxDB v2 CLI](https://docs.influxdata.com/influxdb/v2/tools/influx-cli/).
+7. [Download InfluxDB v2](https://docs.influxdata.com/influxdb/v2/install/). Once you have downloaded InfluxDB v2, make sure the [InfluxDB v2 daemon (`influxd`)](https://docs.influxdata.com/influxdb/v2/reference/cli/influxd/) has been added to your PATH. InfluxDB v2 does not need to be running, the daemon will be used in isolation.
````
will users need to have the daemon running or just added to the path? and before users run the script, how should they check to ensure it's setup correctly? (ie. verify daemon can run from PATH with `$ influxd`?)

trevorbonas replied:
The daemon should just be added to the user's PATH. The script will check and let users know whether it's in their PATH. When deployed in an AMI, user's don't need to worry about where `influxd` is.

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
````diff
@@ -0,0 +1,157 @@
+    - The InfluxDB v2 buckets that you want to migrate and their organizations.
+    - The name of the secret you created in AWS Secrets Manager that contains your InfluxDB v2 and v3 tokens.
+    ```shell
+    python3.14 influxdb_v2_to_v3_migration.py \
````
I get the following error when I try running this script:

```
2025-12-22 11:45:29,423 - influxdb_v2_to_v3_migration - ERROR - Backup failed for bucket testBuk: Command '['influx', 'backup', '--host', 'https://m5li60zx9z-zcipc7m4j2l72o.timestream-influxdb.us-west-2.on.aws:8086', '--compression', 'none', '--org', 'org', '--bucket', 'testBuk', '/Users/fred.park/engine/data']' returned non-zero exit status 1.:
2025/12/22 11:45:29 WARN: Couldn't parse version "dev" reported by server, assuming latest backup/restore APIs are supported
2025/12/22 11:45:29 INFO: Downloading metadata snapshot
Error: failed to backup metadata: failed to download metadata snapshot: 401 Unauthorized: read:authorizations is unauthorized

2025-12-22 11:45:29,465 - influxdb_v2_to_v3_migration - ERROR - Error processing ('testBuk', 'org'): Backup failed for bucket testBuk: Command '['influx', 'backup', '--host', 'https://m5li60zx9z-zcipc7m4j2l72o.timestream-influxdb.us-west-2.on.aws:8086', '--compression', 'none', '--org', 'org', '--bucket', 'testBuk', '/Users/fred.park/engine/data']' returned non-zero exit status 1.:
2025/12/22 11:45:29 WARN: Couldn't parse version "dev" reported by server, assuming latest backup/restore APIs are supported
2025/12/22 11:45:29 INFO: Downloading metadata snapshot
Error: failed to backup metadata: failed to download metadata snapshot: 401 Unauthorized: read:authorizations is unauthorized
```

I've confirmed in the previous step that connectivity is good, but if there are authz requirements can we also confirm/validate this in a prior step?

trevorbonas replied:
You will need to use an operator token. It's the only type of token that has permission `read:authorizations`. I can add a note in the README and link to Timestream's documentation for retrieving the operator token.

My method for getting the operator token is to log into the UI and clone the token.

fredjoonpark replied:
yes let's explicitly mention that they'll need an operator token and show steps or link a guide to how it can be retrieved

trevorbonas replied:
I have added a note that an operator token is required and linked to Timestream's guide for getting an operator token.

fredjoonpark commented:
are we providing any methods for users to validate migrations?

trevorbonas commented:
> are we providing any methods for users to validate migrations?

I'm willing to provide documentation or a script that runs queries for users to verify.

The script reports the number of line protocol points ingested. Users could use this information in combination with `SELECT COUNT(*)` queries to verify, though it would have to be done for each table, with the total for a database equaling the number of points in a bucket's line protocol file.

fredjoonpark commented:
I think we'll at least need a section in the documentation on how users can manually verify, given that the query languages are different across 2 and 3.

Also, will this automation migrate all points up to the time at which it was run (and points written during or after execution will not be considered)? Perhaps a small FAQ section might be good to preempt some of these concerns.

trevorbonas committed:
Update README with instructions for InfluxDB v2 operator tokens

trevorbonas committed:
Add examples for manual verification and use bucket names for LP files

trevorbonas committed:
Add FAQ

trevorbonas commented:
> I think we'll at least need a section in the documentation on how users can manually verify, given that the query languages are different across 2 and 3.
> 
> Also, will this automation migrate all points up to the time at which it was run (and points written during or after execution will not be considered)? Perhaps a small FAQ section might be good to preempt some of these concerns.

I have added a section in the README for manual verification. Output line protocol files now use bucket names to make it easier to do verification.

I have also added a FAQ that answers what the cutoff time is for migrated points among some other questions.

trevorbonas committed:
Reformat FAQ section

trevorbonas committed:
Add --region arguments

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
```diff
@@ -0,0 +1,212 @@
+
+The [Amazon Timestream for InfluxDB](https://docs.aws.amazon.com/timestream/latest/developerguide/timestream-for-influxdb.html) [v2](https://docs.influxdata.com/influxdb/v2/) to [v3](https://docs.influxdata.com/influxdb3/enterprise/) migration script allows you to migrate your data from managed InfluxDB v2 to v3. The script uses the InfluxDB [v2](https://docs.influxdata.com/influxdb/v2/api/v2/) a ...
+
+The script is available standalone or as part of an automated solution that deploys an EC2 instance with the script and all prerequisites installed. If you already have your backed-up data, a separate script that ingests to InfluxDB v3 is provided, [`influxdb_v3_ingestion.py`](./app/influxdb_v3_ingestion.py).
```
I'm not sure how useful it is to link directly to this python script - is there some documentation we can link to instead that talks about how this script can be used? If not, maybe we can do a quick explanation here somewhere. I'd definitely want to know more about this option if my data is already backed up.

trevorbonas replied:
I have added a section in `README.md` with details about how to use the ingestion script, including bash code for doing everything that `influxdb_v2_to_v3_migration.py` does. This link now links to the section.

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
```diff
@@ -0,0 +1,212 @@
+
+### Verifying InfluxDB v3 Count
+
+To check the number of records in a table within a database, use SQL:
```
maybe we can mention here (in brackets) or in the below FAQ that measurements from v2 are tables in v3 (may not be obvious for new v3 users)

trevorbonas replied:
Above, under the `Manual Verification` heading, this is written:
> For verification it is important to note that during the migration process, buckets are mapped to databases, measurements are mapped to tables, and each bucket's data is output to a single line protocol file.

fredjoonpark reviewed (commented):
We now have three folders related to influxdb migrations at `tools/python`:
- influx-migration: migrates between oss idb v2
- influxdb_v2_to_v3_migration: migrates from v2 to v3
- liveanalytics_migration_scripts: migrates from LA to postgres, idb (v2 or v3)

I wonder if we can have a better structure (or even just better naming) for the respective tools (influxdb_v2_to_v3_migration is the clearest) - or at least some top level README that points users interested in influxdb migrations in the right direction. Probably beyond the scope of this PR, but figured I'd raise it here and see what you and @forestmvey think?

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
```diff
@@ -0,0 +1,212 @@
+
+### Steps
+
+1. [Download and install Packer](https://developer.hashicorp.com/packer/install).
```
can we briefly mention why and what Packer and Terraform is used for here?

trevorbonas replied:
Added.

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
````diff
@@ -0,0 +1,212 @@
+   - `runner_ssh_key_name`: The name of an existing key you wish to use to SSH onto your deployed EC2 instance.
+4. Within the [`app`](./app/) directory, initialize Packer and build the AMI, this will produce an AMI in your account with the name `influxdb-v2-to-v3-migration-runner-<timestamp>`:
+   ```shell
+   packer init
````
I get prompted for a required TEMPLATE arg, ie. `packer init packer.pkr.hcl` (I'm on Packer `v1.14.3`)

trevorbonas replied:
Changed to `packer init packer.pkr.hcl`.

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
```diff
@@ -0,0 +1,212 @@
+3. Update [`variables.tf`](./variables.tf), filling in all `"replace me"` placeholders:
+   
+   - `vpc_id`: The ID of an existing VPC.
+   - `subnet_id`: The ID of an existing subnet in the above VPC.
```
does it matter if it's public/private?

trevorbonas replied:
No. Using the pre-built AMI means the VPC doesn't even need an internet gateway.

EDIT: Though downloading and installing Python dependencies may need an internet gateway.

trevorbonas replied:
I have confirmed that with the AMI being setup with `sudo pip install .` if users run the script without a virtual environment, an internet gateway is not necessary.

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
```diff
@@ -0,0 +1,212 @@
+       --cidr <runner_ip>/32
+
+   # InfluxDB v3.
+      aws ec2 authorize-security-group-ingress \
```
nit: extra tab

trevorbonas replied:
Fixed.

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
```diff
@@ -0,0 +1,212 @@
+   - `subnet_id`: The ID of an existing subnet in the above VPC.
+   - `ssh_access_ip`: The IP to grant SSH access to the deployed EC2 instance. For example, `127.0.0.1/32`.
+   - `tokens`: Your InfluxDB v2 and v3 tokens. These tokens will be placed in a secret in AWS Secrets Manager and redacted from all Terraform output.
+   - `runner_ssh_key_name`: The name of an existing key you wish to use to SSH onto your deployed EC2 instance.
```
existing key where? this key needs to exist under EC2 key pairs right? we may want to be more explicit about this and/or add this as a pre-req

trevorbonas replied:
I have now specified that it is an EC2 key pair and have added a step for users to create an EC2 key pair if they don't already have one.

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
````diff
@@ -0,0 +1,212 @@
+   ```
+8. Using the key pair you specified in `variables.tf`, SSH onto the instance, using the output `runner_ip`:
+   ```shell
+   ssh -i <path to key> ec2-user@<runner_ip>
````
```suggestion
   ssh -i <path to private key> ec2-user@<runner_ip>
```

trevorbonas replied:
Added.

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
````diff
@@ -0,0 +1,212 @@
+   terraform init
+   terraform apply
+   ```
+6. Review the proposed changes by Terraform and type `yes`.
````
what are your thoughts on mentioning/explaining what resources will be created from this deployment? personally, I'd like to know what resources terraform will create when following docs without having to look at`main.tf`. if you think it's too pedantic maybe we can just mention a high level description (ie. creates and deploys an ec2 into an existing VPC), or just ask the users to refer to the output from a terraform plan.

trevorbonas replied:
Here (now step 7) I now list all resources `main.tf` will deploy.

trevorbonas commented:
> We now have three folders related to influxdb migrations at `tools/python`:
> 
>     * influx-migration: migrates between oss idb v2
> 
>     * influxdb_v2_to_v3_migration: migrates from v2 to v3
> 
>     * liveanalytics_migration_scripts: migrates from LA to postgres, idb (v2 or v3)
> 
> 
> I wonder if we can have a better structure (or even just better naming) for the respective tools (influxdb_v2_to_v3_migration is the clearest) - or at least some top level README that points users interested in influxdb migrations in the right direction. Probably beyond the scope of this PR, but figured I'd raise it here and see what you and @forestmvey think?

I agree, their naming makes it hard to tell the difference between tools. We could rename the directories to make things clearer, maybe like this:

- `influx-migration` -> `influxdb_v2_backup_restore` or `influxdb_v2_to_v2_migration`.
- `influxdb_v2_to_v3_migration` same.
- `liveanalytics_migration_scripts` -> `live_analytics_multi_destination_migration`.

We've have to make sure to update any blog posts or documentation that links to these directories.

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
````diff
@@ -0,0 +1,212 @@
+5. In the [`influxdb_v2_to_v3_migration`](.) directory, initialize and apply Terraform changes:
+   ```shell
+   terraform init
+   terraform apply
````
I got a bunch of these warnings:
```
==> amazon-ebs.influxdb_v2_to_v3_migration_runner: ./Modules/_decimal/libmpdec/io.c:378:14: note: in expansion of macro ‘EXTRACT_DIGIT’
==> amazon-ebs.influxdb_v2_to_v3_migration_runner:   378 |     case 2:  EXTRACT_DIGIT(s, x, 10UL, dot);
==> amazon-ebs.influxdb_v2_to_v3_migration_runner:       |              ^~~~~~~~~~~~~
==> amazon-ebs.influxdb_v2_to_v3_migration_runner: ./Modules/_decimal/libmpdec/io.c:349:40: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
==> amazon-ebs.influxdb_v2_to_v3_migration_runner:   349 |         if (s == dot) *s++ = '.'; *s++ = '0' + (char)(x / d); x %= d
==> amazon-ebs.influxdb_v2_to_v3_migration_runner:       |                                   ~~~~~^~~~~~~~~~~~~~~~~~~~~
==> amazon-ebs.influxdb_v2_to_v3_migration_runner: ./Modules/_decimal/libmpdec/io.c:378:14: note: in expansion of macro ‘EXTRACT_DIGIT’
==> amazon-ebs.influxdb_v2_to_v3_migration_runner:   378 |     case 2:  EXTRACT_DIGIT(s, x, 10UL, dot);
==> amazon-ebs.influxdb_v2_to_v3_migration_runner:       |              ^~~~~~~~~~~~~
```
should I be concerned?

trevorbonas replied:
No, it is fine. It's also not under our control.

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
````diff
@@ -0,0 +1,212 @@
+4. Within the [`app`](./app/) directory, initialize Packer and build the AMI, this will produce an AMI in your account with the name `influxdb-v2-to-v3-migration-runner-<timestamp>`:
+   ```shell
+   packer init
+   packer build packer.pkr.hcl
````
can we mention here that users can expect this command to take some time (ie. ~15-30 minutes)? for me it took `23 minutes 57 seconds`.

trevorbonas replied:
Added.

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
````diff
@@ -0,0 +1,212 @@
+   ```shell
+   terraform init
+   terraform apply
+   ```
````
I know that users will not need to deploy this more than once, but I get the following errors due to naming conflicts- can we add these to `variables.tf` to make it easily configurable?
```
│ Error: creating Security Group (influxdb-v2-to-v3-migration-security-group): operation error EC2: CreateSecurityGroup, https response error StatusCode: 400, RequestID: 56803757-82e5-4ff0-892c-c0e3bee6de48, api error InvalidGroup.Duplicate: The security group 'influxdb-v2-to-v3-migration-security-group' already exists for VPC 'vpc-zzz'
│ 
│   with aws_security_group.runner_security_group,
│   on main.tf line 24, in resource "aws_security_group" "runner_security_group":
│   24: resource "aws_security_group" "runner_security_group" {
│ 
╵
╷
│ Error: creating Secrets Manager Secret (influxdb_v2_to_v3_migration): operation error Secrets Manager: CreateSecret, https response error StatusCode: 400, RequestID: b694a262-ea93-40eb-9da2-d08c9d61d492, ResourceExistsException: The operation failed because the secret influxdb_v2_to_v3_migration already exists.
│ 
│   with aws_secretsmanager_secret.migration_secret,
│   on main.tf line 65, in resource "aws_secretsmanager_secret" "migration_secret":
│   65: resource "aws_secretsmanager_secret" "migration_secret" {
│ 
╵

│ Error: creating IAM Role (influxdb_v2_to_v3_migration_runner_role): operation error IAM: CreateRole, https response error StatusCode: 409, RequestID: 0343bd18-095e-4b2a-a4e7-05e81aeb91e4, EntityAlreadyExists: Role with name influxdb_v2_to_v3_migration_runner_role already exists.
│ 
│   with aws_iam_role.runner_role,
│   on main.tf line 75, in resource "aws_iam_role" "runner_role":
│   75: resource "aws_iam_role" "runner_role" {
│ 
╵
╷
│ Error: creating IAM Policy (influxdb_v2_to_v3_migration_runner_policy): operation error IAM: CreatePolicy, https response error StatusCode: 409, RequestID: 409624fa-a53b-4c6e-a1ba-6620ed43117e, EntityAlreadyExists: A policy called influxdb_v2_to_v3_migration_runner_policy already exists. Duplicate names are not allowed.
│ 
│   with aws_iam_policy.runner_policy,
│   on main.tf line 93, in resource "aws_iam_policy" "runner_policy":
│   93: resource "aws_iam_policy" "runner_policy" {
│ 

╷
│ Error: creating IAM Instance Profile (influxdb_v2_to_v3_migration_runner_profile): operation error IAM: CreateInstanceProfile, https response error StatusCode: 409, RequestID: 9185378d-4589-44a7-b47e-42fe1f7ff255, EntityAlreadyExists: Instance Profile influxdb_v2_to_v3_migration_runner_profile already exists.
│ 
│   with aws_iam_instance_profile.runner_profile,
│   on main.tf line 121, in resource "aws_iam_instance_profile" "runner_profile":
│  121: resource "aws_iam_instance_profile" "runner_profile" {
│ 

```

fredjoonpark replied:
also -we should mention what IAM permissions users will need (get the following with our dev acc):
```
╷
│ Error: creating IAM Role (influxdb_v2_to_v3_migration_runner_role): operation error IAM: CreateRole, https response error StatusCode: 403, RequestID: 3a733b8c-ca9f-4f3d-b048-bc7d07c13139, api error AccessDenied: User: arn:aws:sts::xxxx:assumed-role/Developer/xxx is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::xxxx:role/influxdb_v2_to_v3_migration_runner_role because no identity-based policy allows the iam:CreateRole action
│ 
│   with aws_iam_role.runner_role,
│   on main.tf line 75, in resource "aws_iam_role" "runner_role":
│   75: resource "aws_iam_role" "runner_role" {
│ 
```

trevorbonas replied:
> I know that users will not need to deploy this more than once, but I get the following errors due to naming conflicts- can we add these to variables.tf to make it easily configurable?

All resources in `main.tf` now use a random prefix. Users shouldn't need to worry about conflicts anymore.

trevorbonas replied:
> also -we should mention what IAM permissions users will need (get the following with our dev acc) . . .

I have added all necessary permissions in the new **Permissions** section. I tested by deploying and deleting all resources with a test role.

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
````diff
@@ -0,0 +1,212 @@
+        --influxdb-v3-url "https://example.com:8181" \
+        --influxdb-v2-buckets-and-orgs "bucket-one:organization-one,bucket-two:organization-two" \
+        --tokens-secret-name "influxdb_v2_to_v3_migration"
+    ```
````
This does not work for me - when I run this I get the following error:
```
2025-12-31 21:56:47,345 - influxdb_v2_to_v3_migration - ERROR - Failed to retrieve tokens from Secrets Manager: An error occurred (AccessDeniedException) when calling the GetSecretValue operation: User: arn:aws:sts::xxxxx:assumed-role/influxdb_v2_to_v3_migration_runner_role-2/i-0c48772511e20aea7 is not authorized to perform: secretsmanager:GetSecretValue on resource: influxdb_v2_mig_test_4 because no identity-based policy allows the secretsmanager:GetSecretValue action
```

trevorbonas replied:
I got this error when using the end of the secret ARN as its name. It seems that secret ARNs include random characters at the end that are not part of the secret's name. Do you think this is what happened for you? I have added a `"secret_name"` output to make this easier.

Unrelated, but also keep in mind that security groups for the source and destination need to be updated after Terraform deployment to allow inbound access from the EC2 runner.

fredjoonpark commented on tools/python/influxdb_v2_to_v3_migration/README.md:
````diff
@@ -0,0 +1,212 @@
+   ```
+   - If you don't want to use SSH, you can use AWS SSM instead:
+      ```shell
+      aws ssm start-session --target <instance ID>
````
An SSM user does not have permissions to EC2 user (cannot cd to `home/ec2-user` where all the migration scripts live) - if we want to support both paths, we should handle this somehow.

trevorbonas replied:
Users will need to switch to the `ec2-user` with:
```shell
sudo su - ec2-user
```

I have added a note that they must do this to the README.

trevorbonas committed:
Add section to README.md about using influxdb_v3_ingestion.py

trevorbonas committed:
Add explanations for why Packer and Terraform are necessary

trevorbonas committed:
Fix packer init command

trevorbonas committed:
Fix indentation in code example

trevorbonas committed:
Add step for creating an EC2 key pair in README

trevorbonas committed:
Specify that the path to the key is private in example command

trevorbonas committed:
Add all resources that will be deployed to README.md

trevorbonas committed:
Add note about Packer run time

trevorbonas committed:
Use random resource prefixes and output secret name

trevorbonas committed:
Add IAM permissions to README.md

trevorbonas committed:
Add SSM session details to README.md

trevorbonas committed:
Change aws_region reference to use region instead of name

forestmvey commented:
> We now have three folders related to influxdb migrations at `tools/python`:
> 
>     * influx-migration: migrates between oss idb v2
> 
>     * influxdb_v2_to_v3_migration: migrates from v2 to v3
> 
>     * liveanalytics_migration_scripts: migrates from LA to postgres, idb (v2 or v3)
> 
> 
> I wonder if we can have a better structure (or even just better naming) for the respective tools (influxdb_v2_to_v3_migration is the clearest) - or at least some top level README that points users interested in influxdb migrations in the right direction. Probably beyond the scope of this PR, but figured I'd raise it here and see what you and @forestmvey think?

I'll make a follow up ticket for this and treat it out of scope for this PR

forestmvey commented on tools/python/influxdb_v2_to_v3_migration/README.md:
```diff
@@ -0,0 +1,432 @@
+    python3.14 influxdb_v2_to_v3_migration.py \
+        --influxdb-v2-url "https://example.com:8086" \
+        --influxdb-v3-url "https://example.com:8181" \
+        --influxdb-v2-buckets-and-orgs "bucket-one:organization-one,bucket-two:organization-two" \
```
It would be good to be able to migrate all buckets within an organization. A parameter `--influxdb-v2-orgs` could be a list of orgs which the script can query for all buckets within the org.

trevorbonas replied:
I have added a `--influxdb-v2-orgs` argument that is mutually exclusive with `--influxdb-v2-buckets-and-orgs`. I have included tests for the new argument.

Additionally, since some buckets in an organization may be empty and exporting empty buckets to line protocol fails, migrations will continue in the event that some of the buckets failed to export to line protocol. This information will be logged for the user to follow up on.

forestmvey commented on tools/python/influxdb_v2_to_v3_migration/app/tests/integration/pytest.ini:
```diff
@@ -0,0 +1,2 @@
+[pytest]
+filterwarnings = ignore::DeprecationWarning:testcontainers.*:
```
It would be better to not use * if possible as this raises eyebrows for anyone looking into the codebase

trevorbonas replied:
I have changed this to be more specific.

forestmvey commented on tools/python/influxdb_v2_to_v3_migration/README.md:
```diff
@@ -0,0 +1,432 @@
+
+## Overview
+
+The [Amazon Timestream for InfluxDB](https://docs.aws.amazon.com/timestream/latest/developerguide/timestream-for-influxdb.html) [v2](https://docs.influxdata.com/influxdb/v2/) to [v3](https://docs.influxdata.com/influxdb3/enterprise/) migration script allows you to migrate your data from managed InfluxDB v2 to v3. The script uses the InfluxDB [v2](https://docs.influxdata.com/influxdb/v2/api/v2/) a ...
```
It may be nicer to have a separate section showing how the migration works; maybe with a diagram. The overview is a mouthful, and I don't think it would be easily understood for how the migration is completed unless someone is very familiar with InfluxDB APIs.

trevorbonas replied:
The introduction is now shorter and is followed by a diagram with a brief explanation of each migration step.

forestmvey commented on tools/python/influxdb_v2_to_v3_migration/README.md:
````diff
@@ -0,0 +1,432 @@
+}
+```
+
+## Standalone Usage
````
I am tempted to split out the automated solution documentation to a nested directory, along with the Terraform files. The top level README can have go over how to use the script standalone, a link to the automated EC2 solution, how the script works (with a diagram?), and a brief FAQ.

trevorbonas replied:
Now, the migration scripts and this README are in the top level directory. The files for automation, `main.tf`, `packer.pkr.hcl`, etc., are in the `automated_deployment` directory. This directory has its own README.

forestmvey commented on tools/python/influxdb_v2_to_v3_migration/automated_deployment/main.tf:
```diff
@@ -0,0 +1,189 @@
+
+resource "aws_instance" "influxdb_v2_to_v3_migration_runner" {
+
+  ami                  = data.aws_ssm_parameter.influxdb_v2_to_v3_migration_runner_ami.value
```
What is the benefit of creating an AMI vs using a script to build python 3.14 from source. Something like the following:

```
sudo yum groupinstall "Development Tools" -y
sudo yum install openssl-devel bzip2-devel libffi-devel -y

version=3.14.2
wget https://www.python.org/ftp/python/${version}/Python-${version}.tgz
tar -xzf Python-${version}.tgz
cd Python-${version}

./configure --enable-optimizations --enable-loadable-sqlite-extensions
make -j$(nproc)
sudo make altinstall

python3.14 --version
```

trevorbonas replied:
The AMI is not just for Python. The AMI has everything required for a migration: the Influx CLI, the InfluxDB v2 daemon, all Python dependencies, and all migration scripts in the default user's home directory.

The AMI is supposed to make deployment easier for users. Also, using the AMI means that migrations can be done without internet access, as all dependencies are pre-installed.

forestmvey commented on tools/python/influxdb_v2_to_v3_migration/README.md:
```diff
@@ -0,0 +1,432 @@
+No, the daemon (`influxd`) simply needs to be in your PATH, available for the script to use.
+
+### Will Python `3.12` work?
+No, you must use Python minimum version `3.14.1`.
```
Is this the only requirement blocking us from removing the AMI. While 3.14.1 allows free threading, is the multi-threading a big enough bottleneck that it warrants requiring customers to create a new AMI. I would assume that we wouldn't see major performance impacts from earlier versions of Python as InfluxDB ingestion will also be a bottleneck.

Requiring version 3.14.1 is a heavy lift that would need justification for the reduced UX, when we can easily install 3.11 via dnf.

trevorbonas replied:
See my above comment. The AMI is not just for Python.

trevorbonas replied:
The project now uses Python 3.13. The AMI will install Python 3.13 from DNF. 

Performance-wise, since the migration script uses `ProcessPoolExecutor` instead of `InterpreterPoolExecutor` anyways, there is no observable performance difference.

Using this Python version and not building from source means building the AMI takes around 5 minutes.

trevorbonas committed:
Make filter for deprecation warning more specific

trevorbonas committed:
Add --influxdb-v2-orgs argument

trevorbonas committed:
Switch to using Python 3.13

trevorbonas committed:
Restructure directories

trevorbonas committed:
Remove boto3-stubs dependency

forestmvey commented on tools/python/influxdb_v2_to_v3_migration/tests/integration/test_influxdb_v2_to_v3_migration.py:
```diff
@@ -0,0 +1,574 @@
+    tokens_secret_name: str = "v2-to-v3-integration-test-secret"
+
+    session: boto3.Session
+    secrets_manager_client: SecretsManagerClient
```
```suggestion
```

trevorbonas replied:
I have changed the type hint to `Any`.

forestmvey commented on tools/python/influxdb_v2_to_v3_migration/README.md:
````diff
@@ -0,0 +1,217 @@
+
+2. Install the optional test dependencies:
+   ```shell
+   python3.14 -m pip install -e '.[test]'
````
```suggestion
   python3.13 -m pip install -e '.[test]'
```

trevorbonas replied:
Changed.

forestmvey commented on tools/python/influxdb_v2_to_v3_migration/README.md:
````diff
@@ -0,0 +1,217 @@
+
+3. Navigate to [`tests/integration/`](./tests/integration/) and run all tests:
+   ```shell
+   python3.14 -m pytest .
````
```suggestion
   python3.13 -m pytest .
```

trevorbonas replied:
Changed.

trevorbonas committed:
Fix Python version in README

trevorbonas committed:
Remove SecretsManagerClient type hint

trevorbonas committed:
Change secrets_manager_client type hint to Any

forestmvey reviewed (approved):
Looks good, TY
